api.js

```diff
@@ -0,0 +1,132 @@
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyServiceGetter(this, "loginManager",
+                                   "@mozilla.org/login-manager;1",
+                                   "nsILoginManager");
+XPCOMUtils.defineLazyModuleGetter(this, "NetUtil",
+                                  "resource://gre/modules/NetUtil.jsm");
+
+const FIELDS = {
+  formSubmitURL: "formSubmitURL",
+  origin: "hostname",
+  realm: "httpRealm",
+  username: "username",
+  password: "password",
+  usernameField: "usernameField",
+  passwordField: "passwordField",
+};
+
+function convert(info) {
+  let obj = {};
+  for (let field of Object.keys(FIELDS)) {
+    obj[field] = info[FIELDS[field]];
+  }
+  return obj;
+}
+
+function match(info, search) {
+  return Object.keys(search).every(field => search[field] == null || search[field] == info[FIELDS[field]]);
+}
+
+function accessible(context, origin) {
+  let url;
+  try {
+    url = NetUtil.newURI(origin);
+  } catch (ex) {
+    // unparseable hostname, can this actually happen?
+    return false;
+  }
+
+  if (url.scheme == "addon") {
+    return (url.path == context.extension.id);
+  } else if (url.scheme == "moz-extension") {
+    return (url.host == context.extension.id
+            || url.host == context.extension.uuid);
+  } else {
+    return (context.extension.whiteListedHosts.matches(url));
+  }
+}
+
+class API extends ExtensionAPI {
+  getAPI(context) {
+    // XXX only return this for background contexts?
+    return {
+      logins: {
+        search(query) {
+          let logins = loginManager.getAllLogins()
+              .filter(login => accessible(context, login.hostname))
+              .filter(login => match(login, query))
+              .map(convert);
+
+          return Promise.resolve(logins);
+        },
+
+        store(info) {
+          let origin = info.origin;
+
+          function check(field) {
+            if (!info[field]) {
+              return;
+            }
+            let uri;
+            try {
+              uri = NetUtil.newURI(info[field]);
+            } catch (err) {
+              return Promise.reject({message: `Cannot parse ${field} as a URL`});
+            }
+
+            if (origin) {
+              if (uri.prePath != origin) {
+                return Promise.reject({message: `Origin does not match ${field}`});
+              }
+            } else {
+              origin = uri.prePath;
+            }
+          }
+
+          check("formSubmitURL");
+          check("realm");
+
+          if (!origin) {
+            return Promise.reject({message: "Must specify origin, formSubmitURL, or realm"});
+          }
+
+          if (!accessible(context, origin)) {
+            return Promise.reject({message: `Permission denied for ${origin}`});
+          }
+
+          let linfo = Cc["@mozilla.org/login-manager/loginInfo;1"]
+              .createInstance(Ci.nsILoginInfo);
```

Comment:
might be nicer to do:
let nsLoginInfo = new Cc("@mozilla.org/login-manager/loginInfo;1", Ci.nsILoginInfo, "init");
let linfo = new nsLoginInfo("http://example.com", "http://example.com", null,
                             "notifyu1", "notifyp1", "user", "pass");